subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/InputChangeAction.java

```diff
@@ -47,7 +47,7 @@ public void execute(InputFileDetails input) {
             javaChangeProcessor.processChange(input, spec);
         } else if (hasExtension(file, ".jar") || hasExtension(file, ".class")) {
             annotationProcessorChangeProcessor.processChange(input, spec);
-        } else {
+        } else if (file.isFile()){
```

Comment:
Nit: missing space between `)` and `{`.

Reply:
❌ What about missing files (i.e files which have been removed)? Should they cause a full recompile? E.g. what happens when a resource has been removed.

With the current change, removing a resource would be ignored.

Same question about other things which may appear in the source directory which are not files (symlinks maybe)?

Reply:
You're right, we'd have distinguish between changes and deletions if we do this. And then we'd still have to recompile on deletions of that doesn't have a known extension. So we're almost back to square one.

Considering this I'm okay with postponing this issue to 5.2 and fixing it differently by only going through this code path when annotation processors are actually present. You can assign that to me.

Reply:
Right, fixing the typo ... and the mistake! Good catch!